nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-components/src/main/java/org/apache/nifi/controller/repository/StandardProcessSession.java

```diff
@@ -356,6 +408,95 @@ private void checkpoint(final boolean copyCollections) {
         checkpoint.checkpoint(this, autoTerminatedEvents, copyCollections);
     }
 
+    private boolean isRetryNeeded(final ProcessorNode processorNode, final StandardRepositoryRecord record, final FlowFileRecord currentFlowFile,
+                                  final int retryCounts, final Map<String, StandardRepositoryRecord> uuidsToRecords) {
+        if (currentFlowFile == null || processorNode == null || processorNode.getRetriedRelationships().isEmpty()) {
+            return false;
+        }
+
+        if (processorNode.isRetriedRelationship(record.getTransferRelationship())) {
+            return retryCounts < processorNode.getRetryCounts();
+        }
+
+        if (forkEventBuilders.get(currentFlowFile) != null) {
+            for (String uuid : forkEventBuilders.get(currentFlowFile).getChildFlowFileIds()) {
+                if (processorNode.isRetriedRelationship(uuidsToRecords.get(uuid).getTransferRelationship())) {
+                    return retryCounts < processorNode.getRetryCounts();
+                }
+            }
+        }
+        return false;
+    }
+
+    private FlowFileRecord updateFlowFileRecord(final StandardRepositoryRecord record,
+                                                final Map<String, StandardRepositoryRecord> uuidsToRecords,
+                                                final int retryCounts, final FlowFileRecord flowFileRecord) {
+
+        removeTemporaryClaim(record);
+        if (forkEventBuilders.get(flowFileRecord) != null) {
+            for (String uuid : forkEventBuilders.get(flowFileRecord).getChildFlowFileIds()) {
+                final StandardRepositoryRecord childRecord = uuidsToRecords.get(uuid);
```

Comment:
Using the uuid map at different places to get records based on their uuid otherwise I would need another loop through records.values to identify the record.